### DISCUSSION TITLE
Declaration of App\Observers\CustomCrawlerObserver::crawled

### DISCUSSION THREAD

**Discussion by linjection** [spatie/crawler#436] [Q&A] [answered]
hi folks,
i always getting this error. i really dont know what i am doing wrong.

Declaration of App\Observers\CustomCrawlerObserver::crawled(Psr\Http\Message\UriInterface $url, Psr\Http\Message\ResponseInterface $response, ?Psr\Http\Message\UriInterface $foundOnUrl = null) must be compatible with Spatie\Crawler\CrawlObservers\CrawlObserver::crawled(Psr\Http\Message\UriInterface $url, Psr\Http\Message\ResponseInterface $response, ?Psr\Http\Message\UriInterface $foundOnUrl = null): void

`

namespace App\Observers;

use DOMDocument;
use Spatie\Crawler\CrawlObservers\CrawlObserver;
use Psr\Http\Message\UriInterface;
use Psr\Http\Message\ResponseInterface;
use GuzzleHttp\Exception\RequestException;
use Illuminate\Support\Facades\Log;

class CustomCrawlerObserver extends CrawlObserver {

    private $content;

    public function __construct() {
        $this->content = NULL;
    }
    /**
     * Called when the crawler will crawl the url.
     *
     * @param UriInterface $url
     */
    public function willCrawl(UriInterface $url): void
    {
        Log::info('willCrawl',['url'=>$url]);
    }

    /**
     * Called when the crawler has crawled the given url successfully.
     *
     * @param UriInterface $url
     * @param ResponseInterface $response
     * @param UriInterface|null $foundOnUrl
     */

    public function crawled(
        UriInterface $url,
        ResponseInterface $response,
        ?UriInterface $foundOnUrl = null
    )
    { /*mycode*/}
`

Anyone a hint?

**Comment by linjection** [comment, marked as answer]
found by myself, missed the : void :)

**Reply by wisefool** [reply]
What does that mean? I'm working with the same example and stuck with the same error.

**Reply by lechdulian** [reply]
@wisefool This means that the return type in function declaration has to be set to void (and the function should not return anything) like so:
```
public function crawled(
        UriInterface $url,
        ResponseInterface $response,
        UriInterface $foundOnUrl = null,
        ?string $linkText = null
    ): void {
        // your function body
    }
    
 ```
 Mind the **void** keyword.

**Comment by Ray-vV** [comment]
> @wisefool This means that the return type in function declaration has to be set to void (and the function should not return anything) like so:
> 
> ```
> public function crawled(
>         UriInterface $url,
>         ResponseInterface $response,
>         UriInterface $foundOnUrl = null,
>         ?string $linkText = null
>     ): void {
>         // your function body
>     }
>     
> ```
> 
> Mind the **void** keyword.


Had the same problem and this wasn't the solution. I noticed most people forget to list half of the parameters the functions need,apparently, which caused my errors.
```
Declaration of App\Observers\CustomCrawlObserver::crawled(Psr\Http\Message\UriInterface $url, Psr\Http\Message\ResponseInterface $response, ?Psr\Http\Message\UriInterface $foundOnUrl = null): void must be compatible with Spatie\Crawler\CrawlObservers\CrawlObserver::crawled(Psr\Http\Message\UriInterface $url, Psr\Http\Message\ResponseInterface $response, ?Psr\Http\Message\UriInterface $foundOnUrl = null, ?string $linkText = null): void
```
Most people only seem to list the Uriinterface $url for the willCrawl() function for some reason? Same for the other functions.
Adding the  ?string $linkText to the willCrawl() and crawled() in my extended custom observer solved this.

TLDR make sure you have got all the same parameters as in the CrawlObserver class.
